File: ruby/strategy/ActionGame/behavior/weapon_behavior.rb
class WeaponBehavior
  def use_weapon
  end
end

class Axe < WeaponBehavior

  def initialize(character_name)
    @character_name = character_name
  end

  def use_weapon
    super
    puts("the " + @character_name + " is fighting with Axe")
  end

  private
  @character_name = nil

end

class Sword < WeaponBehavior

  def initialize(character_name)
    @character_name = character_name
  end

  def use_weapon
    super
    puts("the " + @character_name + " is fighting with Sword")
  end

  private
  @character_name = nil

end

class BowAndArrow < WeaponBehavior

  def initialize(character_name)
    @character_name = character_name
  end

  def use_weapon
    super
    puts("the " + @character_name + " is fighting with BowAndArrow")
  end

  private
  @character_name = nil

end

class Knife < WeaponBehavior

  def initialize(character_name)
    @character_name = character_name
  end

  def use_weapon
    super
    puts("the " + @character_name + " is fighting with Knife")
  end

  private
  @character_name = nil

end

File: javascript/strategy/ActionGame/behavior/WeaponBehavior.js

export class WeaponBehavior {
    useWeapon() {}
}

export class Axe extends WeaponBehavior {

    // private vars
    #character_name

    constructor(character_name) {
        super();
        this.#character_name = character_name
    }

    useWeapon() {
        super.useWeapon();
        console.log("the " + this.#character_name + " is fighting with Axe")
    }
}

export class Sword extends WeaponBehavior {

    // private vars
    #character_name

    constructor(character_name) {
        super();
        this.#character_name = character_name
    }

    useWeapon() {
        super.useWeapon();
        console.log("the " + this.#character_name + " is fighting with Sword")
    }
}

export class BowAndArrow extends WeaponBehavior {

    // private vars
    #character_name

    constructor(character_name) {
        super();
        this.#character_name = character_name
    }

    useWeapon() {
        super.useWeapon();
        console.log("the " + this.#character_name + " is fighting with BowAndArrow")
    }
}

export class Knife extends WeaponBehavior {

    // private vars
    #character_name

    constructor(character_name) {
        super();
        this.#character_name = character_name
    }

    useWeapon() {
        super.useWeapon();
        console.log("the " + this.#character_name + " is fighting with Knife")
    }
}